docs/development/awk/intro-to-awk/intro-to-awk.md

```diff
@@ -0,0 +1,732 @@
+you can execute `hw` as a regular UNIX command:
+
+    ./hw
+{{< output >}}
+Hello World!
+{{< /output >}}
+
+## Variables in AWK
+
+AWK supports two kinds of variables, predefined and user defined ones. Predefined
+variables are defined by AWK itself whereas user defined variables, which are case
+sensitive, are defined by the developer. Notice that AWK automatically initializes
+user defined variables to the **empty string**, which is zero if you convert a variable
+to a number. You are free to convert a string to a number and vice versa as long as
+the string can be converted to a valid number. Last, it is allowed to change the kind
+of value that a variable can hold in your AWK code.
+
+### Predefined Variables
+
+AWK has many built-in variables that are automatically initialized. Some of the
+most important ones are the following:
+
+- The `NF` variable holds the number of fields in the current input record.
+    Notice that each record can have a different number of fields.
+- The `FS` variable is the input field separator. Its default value is the single space
+    but it also matches any sequence of spaces and tabs. Additionally, any number
+	of leading or trailing spaces and tabs is being ignored. If the value of `FS` becomes the null
+	string, then each character in the current line becomes a separate field.
+- The `FILENAME` built-in variable holds the filename of the file currently being
+    processed. Notice that you cannot use `FILENAME` inside the `BEGIN` block because
+	it is not defined yet.
+- The `NR` variable is used for keeping track of the total number of records that have been read so far.
+- The `FNR` variable is used for keeping the total number of records that have been read from the current input file only.
+- The `IGNORECASE` variable is used for telling AWK to ignore case in all of its comparisons or regular expressions.
+- The `ARGC` and `ARGC` variables hold the command line arguments of an AWK program.
+
+### User Defined Variables in AWK
+
+The way you use an AWK variable the first time will define its type. The following
+AWK code uses the `n` variable to count the number of lines that contain a given
+string, which is this case is the word `three`:
+
+    awk '/three/ { n++ }; END { print n+0 }'
+
+The above code says AWK to execute the `n++` code each time there is a match to the
+`/three/` regular expression. When the processing is done, the code in `END` will be
+executed, which just prints the current value of `n` converted to number.
+
+If you provide some input to it, either a filename or user input from the console, you
+will get some output from the AWK code.
+
+## AWK Basics
+
+In this section of the guide you will learn more about the basics of AWK. AWK has support
+for **arrays** and **associative arrays**. It also supports functions and user
+defined functions. AWK programs can have various types of blocks (*patterns*) like
```

Comment:
Can we have a more explicit overview of what constitutes a pattern in AWK? The BEGIN and END and BEGINFILE and ENDFILE sections at line 349 talk about how they are special classes of AWK patterns, but we seem to be lacking a general understanding of how patterns are used in AWK. For comparison, this article lists what things can be patterns:
https://www.gnu.org/software/gawk/manual/html_node/Pattern-Overview.html#Pattern-Overview
And this has a nice description/diagram of how the flow of processing with AWK works:
https://www3.physnet.uni-hamburg.de/physnet/Tru64-Unix/HTML/APS32DTE/WKXXXXXX.HTM

Reply:
Added an explanation of patterns.